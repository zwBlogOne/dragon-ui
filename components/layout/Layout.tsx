import React, { Component, Children, ReactElement } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';

const defaultPrefixCls = 'ui-layout';
class Layout extends Component<PropsType, any> {
  static defaultProps = {
    prefixCls: defaultPrefixCls,
    hasSider: false,
  };

  static Header: any;
  static Sider: any;
  static Footer: any;
  static Content: any;

  render() {
    const { prefixCls, className, children, style } = this.props;

    let hasSider = this.props.hasSider;
    if (!hasSider && prefixCls === defaultPrefixCls) {
      Children.forEach(children, (child) => {
        if ('collapsible' in (child as ReactElement<any>).props) {
          hasSider = true;
        }
      });
    }
    const cls = classnames(prefixCls, {
      [className!]: !!className,
      [`${prefixCls}-has-sider`]: hasSider,
    });

    return (
      <div className={cls} style={style}>
        {children}
      </div>
    );
  }
}

const genSubComponent = (props: PropsType) => {
  return class SubComponent extends Component<PropsType, any> {
    render() {
      const { prefixCls } = props;
      return <Layout prefixCls={prefixCls} {...this.props} />;
    }
  };
};

const Header = genSubComponent({ prefixCls: 'ui-layout-header' });
const Content = genSubComponent({ prefixCls: 'ui-layout-content' });
const Footer = genSubComponent({ prefixCls: 'ui-layout-footer' });

Layout.Header = Header;
Layout.Footer = Footer;
Layout.Content = Content;

export default Layout;
