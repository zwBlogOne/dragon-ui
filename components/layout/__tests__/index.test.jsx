import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Layout from '../index';

const { Header, Content, Footer, Sider } = Layout;

describe('Layout', () => {
  it('renders basic layout correctly', () => {
    const wrapper = render(
      <Layout>
        <Header>Header</Header>
        <Content>Content</Content>
        <Footer>Footer</Footer>
      </Layout>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders basic layout with sider correctly', () => {
    const wrapper = render(
      <Layout>
        <Header>Header</Header>
        <Layout>
          <Sider>Sider</Sider>
          <Content>Content</Content>
        </Layout>
        <Footer>Footer</Footer>
      </Layout>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders sider layout correctly', () => {
    const wrapper = render(
      <Layout>
        <Sider>Sider</Sider>
        <Layout>
          <Header>Header</Header>
          <Content>Content</Content>
        </Layout>
      </Layout>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders collapsible sider layout correctly', () => {
    const wrapper = render(
      <Layout>
        <Sider collapsible>Sider</Sider>
        <Layout>
          <Header>Header</Header>
          <Content>Content</Content>
        </Layout>
      </Layout>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('behave correctly when toggle collapsed status', () => {
    const onCollapse = jest.fn();
    const wrapper = mount(
      <Layout>
        <Sider collapsible onCollapse={onCollapse}>Sider</Sider>
        <Layout>
          <Header>Header</Header>
          <Content>Content</Content>
        </Layout>
      </Layout>
    );
    wrapper.find('.ui-layout-sider-trigger').simulate('click');
    expect(onCollapse).toHaveBeenCalledWith(true);
  });
});
