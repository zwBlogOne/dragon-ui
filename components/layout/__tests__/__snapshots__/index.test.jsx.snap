// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout renders basic layout correctly 1`] = `
<div
  class="ui-layout"
>
  <div
    class="ui-layout-header"
  >
    Header
  </div>
  <div
    class="ui-layout-content"
  >
    Content
  </div>
  <div
    class="ui-layout-footer"
  >
    Footer
  </div>
</div>
`;

exports[`Layout renders basic layout with sider correctly 1`] = `
<div
  class="ui-layout"
>
  <div
    class="ui-layout-header"
  >
    Header
  </div>
  <div
    class="ui-layout ui-layout-has-sider"
  >
    <div
      class="ui-layout-sider"
      style="flex:0 0 236;max-width:236px;min-width:236px;width:236px"
    >
      <div
        class="ui-layout-sider-children"
      >
        Sider
      </div>
    </div>
    <div
      class="ui-layout-content"
    >
      Content
    </div>
  </div>
  <div
    class="ui-layout-footer"
  >
    Footer
  </div>
</div>
`;

exports[`Layout renders collapsible sider layout correctly 1`] = `
<div
  class="ui-layout ui-layout-has-sider"
>
  <div
    class="ui-layout-sider"
    style="flex:0 0 236;max-width:236px;min-width:236px;width:236px"
  >
    <div
      class="ui-layout-sider-children"
    >
      Sider
    </div>
    <div
      class="ui-layout-sider-trigger"
      style="width:236px"
    >
      <i
        class="ui-icon ui-icon-arrow-left theme-default"
      />
    </div>
  </div>
  <div
    class="ui-layout"
  >
    <div
      class="ui-layout-header"
    >
      Header
    </div>
    <div
      class="ui-layout-content"
    >
      Content
    </div>
  </div>
</div>
`;

exports[`Layout renders sider layout correctly 1`] = `
<div
  class="ui-layout ui-layout-has-sider"
>
  <div
    class="ui-layout-sider"
    style="flex:0 0 236;max-width:236px;min-width:236px;width:236px"
  >
    <div
      class="ui-layout-sider-children"
    >
      Sider
    </div>
  </div>
  <div
    class="ui-layout"
  >
    <div
      class="ui-layout-header"
    >
      Header
    </div>
    <div
      class="ui-layout-content"
    >
      Content
    </div>
  </div>
</div>
`;
